File: src/main/java/io/github/suitetecsa/sdk/nauta/utils/HtmlErrorParser.java
package io.github.suitetecsa.sdk.nauta.utils;

import org.jetbrains.annotations.NotNull;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class HtmlErrorParser {
    private PortalManager portalManager = PortalManager.USER;
    private Pattern getRegex() {
        return switch (portalManager) {
            case CONNECT -> Pattern.compile("alert\\(\"(?<reason>[^\"]*?)\"\\)");
            case USER -> Pattern.compile("toastr\\.error\\('(?<reason>.*)'\\)");
        };
    }

    public HtmlErrorParser whenPortalManager(PortalManager portalManager) {
        this.portalManager = portalManager;
        return this;
    }

    /**
     * Parses the error message from the given HTML document.
     *
     * @param document The HTML document to parse.
     * @return The parsed error message, or null if no error message is found.
     */
    public String parseError(@NotNull Document document) {
        Elements scripts = document.select("script[type='text/javascript']");
        Element lastScript = scripts.last();

        if (lastScript != null) {
            String scriptData = lastScript.data().trim();
            Matcher matcher = getRegex().matcher(scriptData);

            if (matcher.find()) {
                String reason = matcher.group("reason");
                if (reason != null) {
                    Element errorMsg = Jsoup.parse(reason).selectFirst("li[class=\"msg_error\"]");
                    if (errorMsg != null) {
                        return errorMsg.text();
                    }
                }
            }
        }

        return null;
    }
}

File: src/main/kotlin/io/github/suitetecsa/sdk/nauta/utils/HtmlErrorParser.kt
package io.github.suitetecsa.sdk.nauta.utils

import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import java.util.regex.Pattern

object HtmlErrorParser {
    private val regex: Pattern
        get() = Pattern.compile("alert\\(\"(?<reason>[^\"]*?)\"\\)")

    /**
     * Parses the error message from the given HTML document.
     *
     * @param document The HTML document to parse.
     * @return The parsed error message, or null if no error message is found.
     */
    @JvmStatic
    fun parseError(document: Document): String? {
        val scripts = document.select("script[type='text/javascript']")
        val lastScript = scripts.last()

        if (lastScript != null) {
            val scriptData = lastScript.data().trim { it <= ' ' }
            val matcher = regex.matcher(scriptData)

            if (matcher.find()) {
                val reason = matcher.group("reason")
                if (reason != null) {
                    val errorMsg = Jsoup.parse(reason).selectFirst("li[class=\"msg_error\"]")
                    if (errorMsg != null) {
                        return errorMsg.text()
                    }
                }
            }
        }

        return null
    }
}
